### PR TITLE
Allow stopwatch and printcounter to go over 18:12:15

### PR DESCRIPTION
**Pull request by thinkyhead**
Addressing #4195

After 18 hours, 12 minutes, and 16 seconds, a 16-bit integer counting seconds will overflow. This PR alters variables in stopwatch and printcounter to use `millis_t` instead of 16-bit integers.
- Use wider integers to store time durations
- Extend `M31` to include days in its output
- Run `M31` at the end of an SD print rather than custom-printing the time

### CODE REVIEW

**Comment by thinkyhead** [comment]
@MarlinFirmware/host-software-team 
Do hosts expect the output of `M31` to never include hours (only min and sec)?

**Comment by foosel** [comment]
Stock OctoPrint doesn't even parse M31 output, so wouldn't be affected by any changes there. Don't know about other hosts though.

**Comment by petrzjunior** [comment]
@thinkyhead Oh, good idea to prevent overflow! I was trying to do exactly the same yesterday, but I wasn't successful enough to push it.

I'm not sure, what `millis_t`  is, but I was thinking about `uint32_t`. This would extend it to 4 billion second, roughly 50 000 days. I was also trying to reformat the info menu to show the timer on new line as "`_days_`d `_hours_`h `_minutes_`m", this would with 50 thousand days fit perfectly even into 14 chars on LCD.

But strange things happened and this snippet

``` cpp
sprintf(printTime, "%02ud %02uh %02um", 
        (uint16_t)(stats.printTime / 86400), 
        ((uint32_t)(stats.printTime / 60)) % 60, 
        ((uint32_t)(stats.printTime / 3600)) % 24
);
```

doesn't work on Arduino while in [online C++ Shell](http://cpp.sh/) does.
It skips the middle (sometimes the last) value while formatting. Strange. Blindness or Arduino bug?

I would certainly implement days, since @clexpert's printer is printing more than someone would expect. :smile:

**Comment by thinkyhead** [comment]
> I'm not sure, what `millis_t` is…

A search of the source code reveals…

``` cpp
typedef unsigned long millis_t;
```

> strange things happened and this snippet

I would suggest instead…

``` cpp
sprintf(printTime, "%id %02ih %02im", 
        int(stats.printTime / 60 / 60 / 24),
        int(stats.printTime / 60 / 60) % 60,
        int(stats.printTime / 60) % 60
);
```

File: Marlin/language_de.h [hunk 1]
```diff
@@ -179,9 +179,16 @@
 #define MSG_INFO_EXTRUDERS                  "Extruders"
 #define MSG_INFO_BAUDRATE                   "Baud"
 #define MSG_INFO_PROTOCOL                   "Protokol"
-#define MSG_INFO_TOTAL_PRINTS               "Gesamte Drucke"
-#define MSG_INFO_FINISHED_PRINTS            "Beendete Drucke"
-#define MSG_INFO_PRINT_TIME                 "Gesamte Druckzeit"
+
+#if LCD_WIDTH > 19
+  #define MSG_INFO_TOTAL_PRINTS             "Gesamte Drucke   "
```
Comment (ghost) [inline Marlin/language_de.h:184]: It seems that `MSG_INFO_PRINT_COUNT` is right.

**Comment by ghost** [comment]
Excuse me I've a question, are `MSG_END_DAY`, `MSG_END_HOUR`, and `MSG_END_MINUTE` used where?

**Comment by jbrazio** [comment]
I can't find their usage anywhere in the code (current RCBugfix).. maybe on a unmerged PR ?

**Comment by Blue-Marlin** [comment]
Yesterday removed with https://github.com/MarlinFirmware/Marlin/commit/34da77dad3873ff75f447465d4cca45a464db9cd

**Comment by thinkyhead** [comment]
Hmm, guess those strings are now obsolete. Print time on the display is now coming from M31.

**Comment by ghost** [comment]
I see, thus I'll remove those from language files that I'm managing.

**Comment by kakaroto** [comment]
a bit late here, but just to confirm that Cura LE also doesn't use M31 and counts the print time internally (and stops/resumes the counter when user pauses/resumes the print).

**Pull request merged**